### DISCUSSION TITLE
Combined race scripts

### DISCUSSION THREAD

**Discussion by Zombie-Ryushu** [jncraton/BWMetaAI#61] [General]
Is there a way to program the scripts that if they have another race's buildings. (Protoss with Terran Buildings included) The Protoss could do combined attack strategies leveraging two forces?

**Comment by jncraton** [comment]
This may be possible generally, but this project focuses on melee mode, where each race will start with just the town hall and 4 workers.

This project is freely licensed, so you are welcome to use it as the basis for scripts that meet your particular needs if that is helpful to you.

**Comment by Zombie-Ryushu** [comment]
I've been working on Hybrid Scripts for Raynor's Raiders. 
I'm looking to try and create a Joint Terran and Protoss Script using BWMetaAI to simulate a battle against Raynor's Raiders using Strategies a Human player would.
Such as Zealot waves with Terran Medics supporting.
Battle Cruisers Cloaked by Arbiters and Yamato Guns. I've managed to get this partially working, but there are situations where the AI "Just stops"

**Comment by jncraton** [comment]
That makes sense. I've not done what you are doing where a single script is controlling units from multiple factions in a campaign. It might be easier to start from a blank script using [PyAI](https://github.com/poiuyqwert/PyMS#pyai) or a similar tool rather than attempting to adapt BWMetaAI for this purpose.

**Comment by Zombie-Ryushu** [comment]
I have had some limited success with this idea. I can get the Protoss faction I programmed to Build Goliaths, Marines, and Wraiths. My latest headache has been the Protoss Supply Blocking themselves by not realizing they can build new Terran Supply Depots.

If a Script telling them they should build 8 Dragoons and 8 Goliaths gets supply blocked because they don't know they can build new Supply Depots they will stop producing Goliaths and Dragoons as well.

**Comment by Zombie-Ryushu** [comment]
More progress. The Protoss will research more stuff, but they are slow and inefficient. I'd like you to look at my code.
Protoss will:
Build Terran Barracks, Engineering Bay, Star Port, and one supply Depot.
They will Upgrade infantry weapons.
They will train Marines, Vultures, Goliaths, Wraiths, and Medics but not Siege Tanks, or BattleCruisers
They will Send attack waves consisting of Dragoons and Marines, Goliaths, and Dragoons both supported by medics, but they are slow and inefficient. They take too long.

**Comment by Zombie-Ryushu** [comment]
I created a fork. I have a crash I am trying to Diagnose.

**Comment by Zombie-Ryushu** [comment]
I created my own fork I'd like you to look at.

**Comment by Zombie-Ryushu** [comment]
I have a mostly functional version of this, but it overwrites default Protoss behavior. It's also still too slow for the other AIs in many cases, but it works. I'd still like you to look at the code and absorb it as a bespoke commit.

**Reply by jncraton** [reply]
I'll echo what I said previously:

> this project focuses on melee mode, where each race will start with just the town hall and 4 workers.

I don't see a reason to attempt to integrate your separate project back into this one.

**Comment by Zombie-Ryushu** [comment]
That is still true of this fork. My test maps generally consist of one Terran Command Center and one Protoss Nexus with four SCVs and Four Probes next door to each other.

**Reply by jncraton** [reply]
Perhaps I misunderstood. I was under the impression that the Nexus and CC were both controlled by the same player, which is very different than regular melee.

**Comment by Zombie-Ryushu** [comment]
They are. The setup runs in "Use Map Settings"
If I use melee mode it will deprive the Protoss player of a Command Center. The intended setup is 'simulated Melee' Where the circumstances of Melee exist, and the player is intended to build everything from a Town Hall up.

**Reply by jncraton** [reply]
That makes sense. I'm glad that you were able to get that working, but it isn't something that I'm going to merge back into this project, as it focuses on emulating the native multiplayer melee experience.
